src/query/list_file.c

```diff
@@ -392,6 +394,40 @@ qfile_list_cache_cleanup (THREAD_ENTRY * thread_p)
 }
 #endif
 
+int
+qcache_get_new_ht_no (THREAD_ENTRY * thread_p)
+{
+  int ht_no = -1;
+
+  if (qfile_List_cache.next_ht_no >= 0)
+    {
+      ht_no = qfile_List_cache.next_ht_no;
+      qfile_List_cache.next_ht_no = qfile_List_cache.free_ht_list[qfile_List_cache.next_ht_no];
+    }
+
+  return ht_no;
+}
+
+void
+qcache_free_ht_no (THREAD_ENTRY * thread_p, int ht_no)
+{
+  if (QFILE_IS_LIST_CACHE_DISABLED || ht_no < 0)
```

Comment:
this free routine is called from only qfile_end_use_of_list_cache_entry, so critical section is not necessary. I'm going to remove the critical section.